File: macos/UniviewCodeBuddyProxy/Sources/Services/ResponsesStreamTranslator.swift
import Foundation

// ═══════════════════════════════════════════════════════════════
// ResponsesStreamTranslator — OpenAI → Responses API SSE 状态机
// 逐块将上游 OpenAI Chat Completions SSE 转换为 Responses API SSE
// ═══════════════════════════════════════════════════════════════

final class ResponsesStreamTranslator {

    // MARK: - Tool Call State

    private struct ToolCallState {
        var id: String = ""
        var name: String = ""
        var arguments: String = ""
        var outputIndex: Int = 0
        var started: Bool = false
    }

    // MARK: - State

    /// 是否已发送 response.created + response.in_progress
    private var started: Bool = false
    /// 是否已发送 response.completed
    private var finished: Bool = false
    /// 文本输出是否已开始
    private var textStarted: Bool = false
    /// 当前 content index（用于 output_text 事件）
    private var contentIndex: Int = 0
    /// prompt token 计数
    private var promptTokens: Int = 0
    /// completion token 计数
    private var completionTokens: Int = 0
    /// finish reason
    private var finishReason: String? = nil
    /// 累积的完整文本内容
    private var fullContent: String = ""
    /// tool_call 状态跟踪（按 OpenAI tool_call index）
    private var toolCalls: [Int: ToolCallState] = [:]
    /// 维护 tool_call 的出现顺序
    private var toolCallOrder: [Int] = []

    /// 客户端请求的模型名
    private let requestedModel: String
    /// response ID
    private let respID: String
    /// 文本输出项 ID
    private let outputItemID: String

    // MARK: - Init

    init(requestedModel: String) {
        self.requestedModel = requestedModel
        self.respID = "resp_" + Self.randomHex(24)
        self.outputItemID = "msg_" + Self.randomHex(24)
    }

    // MARK: - Process

    /// 处理一行上游 SSE 数据，返回零条或多条 Responses API SSE 事件字符串
    func process(line: String) -> [String] {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return [] }

        let (dataStr, done, ok) = Self.parseSSELine(trimmed)
        guard ok else { return [] }

        if done {
            if !finished {
                finished = true
                return emitFinish()
            }
            return []
        }

        guard let data = dataStr.data(using: .utf8),
              let chunk = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return []
        }

        // 提取 usage
        if let usage = chunk["usage"] as? [String: Any] {
            if let pt = usage["prompt_tokens"] as? Int { promptTokens = pt }
            else if let pt = usage["prompt_tokens"] as? Double { promptTokens = Int(pt) }

            if let ct = usage["completion_tokens"] as? Int { completionTokens = ct }
            else if let ct = usage["completion_tokens"] as? Double { completionTokens = Int(ct) }
        }

        var events: [String] = []

        guard let choices = chunk["choices"] as? [[String: Any]] else { return [] }

        for choice in choices {
            guard let delta = choice["delta"] as? [String: Any] else { continue }

            // 首次收到有效内容 → 发送 response.created + response.in_progress
            if !started {
                started = true
                events.append(contentsOf: emitResponseCreated())
            }

            // ── content（text）──

            if let content = delta["content"] as? String, !content.isEmpty {
                events.append(contentsOf: emitTextStartIfNeeded())
                fullContent += content
                events.append(responsesSSE("response.output_text.delta", [
                    "type": "response.output_text.delta",
                    "output_index": 0,
                    "content_index": contentIndex,
                    "delta": content,
                ]))
            }

            // ── tool_calls ──

            if let toolCallDeltas = delta["tool_calls"] as? [[String: Any]] {
                for tcMap in toolCallDeltas {
                    let tcIdx = Self.intFromMap(tcMap, key: "index")

                    // 首次出现此 tool_call 索引
                    if toolCalls[tcIdx] == nil {
                        toolCalls[tcIdx] = ToolCallState()
                        toolCallOrder.append(tcIdx)
                    }

                    var tcState = toolCalls[tcIdx]!

                    // 有 ID → 首次出现此 tool_call
                    if let id = tcMap["id"] as? String, !id.isEmpty {
                        tcState.id = id
                        if !tcState.started {
                            tcState.started = true
                            var fnName = ""
                            if let fn = tcMap["function"] as? [String: Any] {
                                fnName = fn["name"] as? String ?? ""
                                tcState.name = fnName
                            }

                            let outputIdx = computeToolOutputIndex(tcIdx: tcIdx)

                            // function_call 输出项添加
                            events.append(responsesSSE("response.output_item.added", [
                                "type": "response.output_item.added",
                                "output_index": outputIdx,
                                "item": [
                                    "type": "function_call",
                                    "id": id,
                                    "call_id": id,
                                    "name": fnName,
                                    "status": "in_progress",
                                ],
                            ]))
                        }
                    }

                    // function arguments delta
                    if let fn = tcMap["function"] as? [String: Any] {
                        if let args = fn["arguments"] as? String, !args.isEmpty {
                            tcState.arguments += args
                            // 仅在 output_item.added 已发送后才发送 arguments delta
                            if tcState.started {
                                let outputIdx = computeToolOutputIndex(tcIdx: tcIdx)
                                events.append(responsesSSE("response.function_call_arguments.delta", [
                                    "type": "response.function_call_arguments.delta",
                                    "output_index": outputIdx,
                                    "item_id": tcState.id,
                                    "delta": args,
                                ]))
                            }
                        }
                        // 也提取 name（可能在后续 chunk 中出现）
                        if let name = fn["name"] as? String, !name.isEmpty, tcState.name.isEmpty {
                            tcState.name = name
                        }
                    }

                    toolCalls[tcIdx] = tcState
                }
            }

            // ── finish_reason ──

            if let fr = choice["finish_reason"] as? String, !fr.isEmpty, !finished {
                finishReason = fr
            }
        }

        return events
    }

    // MARK: - Close (for stream end / errors)

    /// 关闭流（在流结束或出错时调用），返回剩余事件
    func close() -> [String] {
        if finished { return [] }
        finished = true

        var events: [String] = []

        if started {
            events.append(contentsOf: closeTextBlock())
            events.append(contentsOf: closeToolCallBlocks())
        } else {
            // 未收到任何内容就出错 → 先发送 response.created
            events.append(contentsOf: emitResponseCreated())
        }

        // truncated completed
        events.append(responsesSSE("response.completed", [
            "type": "response.completed",
            "response": buildResponsesSSEObject(status: "incomplete", outputItems: nil),
        ]))

        return events
    }

    // MARK: - Error Output

    /// 生成 Responses API 格式 SSE 错误
    static func sseError(message: String) -> String {
        let msg = message
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
        return "event: error\ndata: {\"type\":\"error\",\"error\":{\"type\":\"server_error\",\"message\":\"\(msg)\"}}\n\n"
    }

    /// 生成 Responses API 格式 SSE 上下文超限错误
    static func sseContextLimitError(message: String) -> String {
        let msg = ("request too large: " + message)
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
        return "event: error\ndata: {\"type\":\"error\",\"error\":{\"type\":\"invalid_request_error\",\"message\":\"\(msg)\"}}\n\n"
    }

    // MARK: - Private Event Emitters

    /// 发送 response.created + response.in_progress
    private func emitResponseCreated() -> [String] {
        let createdResp = buildResponsesSSEObject(status: "in_progress", outputItems: nil)
        return [
            responsesSSE("response.created", [
                "type": "response.created",
                "response": createdResp,
            ]),
            responsesSSE("response.in_progress", [
                "type": "response.in_progress",
                "response": createdResp,
            ]),
        ]
    }

    /// 首次有文本内容时发送文本输出开始事件
    private func emitTextStartIfNeeded() -> [String] {
        guard !textStarted else { return [] }
        textStarted = true

        return [
            // 输出项添加
            responsesSSE("response.output_item.added", [
                "type": "response.output_item.added",
                "output_index": 0,
                "item": [
                    "type": "message",
                    "id": outputItemID,
                    "status": "in_progress",
                    "role": "assistant",
                    "content": [] as [Any],
                ],
            ]),
            // 内容部分添加
            responsesSSE("response.content_part.added", [
                "type": "response.content_part.added",
                "output_index": 0,
                "content_index": contentIndex,
                "part": [
                    "type": "output_text",
                    "text": "",
                ],
            ]),
        ]
    }

    /// 流正常结束：关闭文本块 + 工具块 + 发送 completed
    private func emitFinish() -> [String] {
        var events: [String] = []
        events.append(contentsOf: closeTextBlock())
        events.append(contentsOf: closeToolCallBlocks())
        events.append(contentsOf: emitCompleted())
        return events
    }

    /// 关闭文本输出块
    private func closeTextBlock() -> [String] {
        guard textStarted else { return [] }

        return [
            // output_text.done
            responsesSSE("response.output_text.done", [
                "type": "response.output_text.done",
                "output_index": 0,
                "content_index": contentIndex,
                "text": fullContent,
            ]),
            // content_part.done
            responsesSSE("response.content_part.done", [
                "type": "response.content_part.done",
                "output_index": 0,
                "content_index": contentIndex,
                "part": [
                    "type": "output_text",
                    "text": fullContent,
                    "annotations": [] as [Any],
                ],
            ]),
            // output_item.done
            responsesSSE("response.output_item.done", [
                "type": "response.output_item.done",
                "output_index": 0,
                "item": [
                    "type": "message",
                    "id": outputItemID,
                    "status": "completed",
                    "role": "assistant",
                    "content": [
                        [
                            "type": "output_text",
                            "text": fullContent,
                            "annotations": [] as [Any],
                        ],
                    ],
                ],
            ]),
        ]
    }

    /// 关闭所有 tool_call 输出块
    private func closeToolCallBlocks() -> [String] {
        var events: [String] = []

        for tcIdx in toolCallOrder {
            guard let tc = toolCalls[tcIdx], tc.started else { continue }
            let outputIdx = computeToolOutputIndex(tcIdx: tcIdx)

            // function_call_arguments.done
            events.append(responsesSSE("response.function_call_arguments.done", [
                "type": "response.function_call_arguments.done",
                "output_index": outputIdx,
                "item_id": tc.id,
                "arguments": tc.arguments,
            ]))
            // output_item.done
            events.append(responsesSSE("response.output_item.done", [
                "type": "response.output_item.done",
                "output_index": outputIdx,
                "item": [
                    "type": "function_call",
                    "id": tc.id,
                    "call_id": tc.id,
                    "name": tc.name,
                    "arguments": tc.arguments,
                    "status": "completed",
                ],
            ]))
        }

        return events
    }

    /// 发送 response.completed 事件
    private func emitCompleted() -> [String] {
        let status = (finishReason == "length") ? "incomplete" : "completed"

        // 构建输出项列表
        var outputItems: [[String: Any]] = []

        if textStarted {
            outputItems.append([
                "type": "message",
                "id": outputItemID,
                "status": "completed",
                "role": "assistant",
                "content": [
                    [
                        "type": "output_text",
                        "text": fullContent,
                        "annotations": [] as [Any],
                    ],
                ],
            ])
        }

        for tcIdx in toolCallOrder {
            guard let tc = toolCalls[tcIdx] else { continue }
            outputItems.append([
                "type": "function_call",
                "id": tc.id,
                "call_id": tc.id,
                "name": tc.name,
                "arguments": tc.arguments,
                "status": "completed",
            ])
        }

        let completedResp = buildResponsesSSEObject(status: status, outputItems: outputItems)
        return [responsesSSE("response.completed", [
            "type": "response.completed",
            "response": completedResp,
        ])]
    }

    // MARK: - Private Helpers

    /// 计算 tool_call 的 output_index
    /// text 输出占 index 0，tool_call 从 textStarted 后开始偏移
    private func computeToolOutputIndex(tcIdx: Int) -> Int {
        let base = textStarted ? 1 : 0
        return base + Self.indexOfInt(toolCallOrder, tcIdx)
    }

    /// 构建用于 SSE 事件的 response 对象
    private func buildResponsesSSEObject(status: String, outputItems: [[String: Any]]?) -> [String: Any] {
        return [
            "id": respID,
            "object": "response",
            "created_at": Int(Date().timeIntervalSince1970),
            "model": requestedModel,
            "status": status,
            "output": outputItems ?? [] as [Any],
            "usage": [
                "input_tokens": promptTokens,
                "output_tokens": completionTokens,
                "total_tokens": promptTokens + completionTokens,
            ],
            "metadata": [:] as [String: Any],
        ]
    }

    /// 生成 Responses API SSE 事件字符串
    private func responsesSSE(_ event: String, _ data: [String: Any]) -> String {
        guard let jsonData = try? JSONSerialization.data(withJSONObject: data),
              let jsonStr = String(data: jsonData, encoding: .utf8) else {
            return ""
        }
        return "event: \(event)\ndata: \(jsonStr)\n\n"
    }

    /// 解析 SSE 行
    private static func parseSSELine(_ line: String) -> (data: String, done: Bool, ok: Bool) {
        var dataStr = line
        if line.hasPrefix("data: ") {
            dataStr = String(line.dropFirst(6))
        } else if line.hasPrefix("data:") {
            dataStr = String(line.dropFirst(5))
        } else {
            return ("", false, false)
        }

        dataStr = dataStr.trimmingCharacters(in: .whitespaces)

        if dataStr == "[DONE]" {
            return ("", true, true)
        }
        if dataStr.isEmpty {
            return ("", false, false)
        }

        return (dataStr, false, true)
    }

    /// 从字典中提取整数值
    private static func intFromMap(_ map: [String: Any], key: String) -> Int {
        if let v = map[key] as? Int { return v }
        if let v = map[key] as? Double { return Int(v) }
        return 0
    }

    /// 返回值在数组中的索引，不存在返回 -1
    private static func indexOfInt(_ slice: [Int], _ val: Int) -> Int {
        for (i, v) in slice.enumerated() {
            if v == val { return i }
        }
        return -1
    }

    /// 生成随机十六进制字符串
    private static func randomHex(_ n: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: n)
        _ = SecRandomCopyBytes(kSecRandomDefault, n, &bytes)
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}

File: windows/UniviewCodeBuddyProxy.Core/Services/ResponsesStreamTranslator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;

namespace UniviewCodeBuddyProxy.Services;

/// <summary>
/// OpenAI to Responses API SSE state machine translator.
/// Emits response.created/in_progress/output_item.added/content_part.added/delta/done events.
/// </summary>
public sealed class ResponsesStreamTranslator
{
    // ═══ Tool Call State ═══

    private sealed class ToolCallState
    {
        public string ID { get; set; } = "";
        public string Name { get; set; } = "";
        public string Arguments { get; set; } = "";
        public int OutputIndex { get; set; }
        public bool Started { get; set; }
    }

    // ═══ State ═══

    private bool _started;
    private bool _finished;
    private bool _textStarted;
    private int _contentIndex;
    private int _promptTokens;
    private int _completionTokens;
    private int _totalTokens;
    private int _cacheReadInputTokens;
    private int _cacheCreationInputTokens;
    private double _credit;
    private string? _finishReason;
    private string _fullContent = "";
    private readonly Dictionary<int, ToolCallState> _toolCalls = new();
    private readonly List<int> _toolCallOrder = new();

    private readonly string _requestedModel;
    private readonly string _respID;
    private readonly string _outputItemID;

    // ═══ Constructor ═══

    public ResponsesStreamTranslator(string requestedModel)
    {
        _requestedModel = requestedModel;
        _respID = "resp_" + RandomHex(24);
        _outputItemID = "msg_" + RandomHex(24);
    }

    // ═══ Process ═══

    /// <summary>
    /// Process a single upstream SSE line, returning zero or more Responses API SSE event strings.
    /// </summary>
    public List<string> ProcessLine(string line)
    {
        var trimmed = line.Trim();
        if (string.IsNullOrEmpty(trimmed)) return [];

        var (dataStr, done, ok) = ParseSSELine(trimmed);
        if (!ok) return [];

        if (done)
        {
            if (!_finished)
            {
                _finished = true;
                return EmitFinish();
            }
            return [];
        }

        JsonElement chunk;
        try
        {
            chunk = JsonDocument.Parse(dataStr).RootElement;
        }
        catch
        {
            return [];
        }

        // Extract usage
        if (chunk.TryGetProperty("usage", out var usage))
        {
            if (usage.TryGetProperty("prompt_tokens", out var pt))
                _promptTokens = GetInt(pt);
            if (usage.TryGetProperty("completion_tokens", out var ct))
                _completionTokens = GetInt(ct);
            if (usage.TryGetProperty("total_tokens", out var tt) && tt.ValueKind == JsonValueKind.Number)
                _totalTokens = tt.GetInt32();
            if (usage.TryGetProperty("credit", out var cr) && cr.ValueKind == JsonValueKind.Number)
                _credit = cr.GetDouble();
            if (usage.TryGetProperty("prompt_tokens_details", out var ptd))
            {
                if (ptd.TryGetProperty("cached_tokens", out var cachedEl))
                {
                    var v = GetInt(cachedEl);
                    if (v > 0) _cacheReadInputTokens = v;
                }
            }
            if (usage.TryGetProperty("prompt_cache_hit_tokens", out var hit) && hit.ValueKind == JsonValueKind.Number)
            {
                var v = hit.GetInt32();
                if (v > 0) _cacheReadInputTokens = v;
            }
            if (usage.TryGetProperty("prompt_cache_miss_tokens", out var miss) && miss.ValueKind == JsonValueKind.Number)
            {
                var v = miss.GetInt32();
                if (v > 0) _cacheCreationInputTokens = v;
            }
            if (usage.TryGetProperty("prompt_cache_write_tokens", out var write) && write.ValueKind == JsonValueKind.Number)
            {
                var v = write.GetInt32();
                if (v > 0) _cacheCreationInputTokens = v;
            }
        }

        var events = new List<string>();

        if (!chunk.TryGetProperty("choices", out var choices) || choices.ValueKind != JsonValueKind.Array)
            return [];

        foreach (var choice in choices.EnumerateArray())
        {
            if (!choice.TryGetProperty("delta", out var delta)) continue;

            // First content -> send response.created + response.in_progress
            if (!_started)
            {
                _started = true;
                events.AddRange(EmitResponseCreated());
            }

            // ── content (text) ──

            if (delta.TryGetProperty("content", out var contentEl))
            {
                var content = contentEl.GetString() ?? "";
                if (!string.IsNullOrEmpty(content))
                {
                    events.AddRange(EmitTextStartIfNeeded());
                    _fullContent += content;
                    events.Add(ResponsesSSE("response.output_text.delta", new Dictionary<string, object>
                    {
                        ["type"] = "response.output_text.delta",
                        ["output_index"] = 0,
                        ["content_index"] = _contentIndex,
                        ["delta"] = content,
                    }));
                }
            }

            // ── tool_calls ──

            if (delta.TryGetProperty("tool_calls", out var toolCallsEl) && toolCallsEl.ValueKind == JsonValueKind.Array)
            {
                foreach (var tcMap in toolCallsEl.EnumerateArray())
                {
                    var tcIdx = IntFromMap(tcMap, "index");

                    if (!_toolCalls.ContainsKey(tcIdx))
                    {
                        _toolCalls[tcIdx] = new ToolCallState();
                        _toolCallOrder.Add(tcIdx);
                    }

                    var tcState = _toolCalls[tcIdx];

                    // Has ID -> first appearance of this tool_call
                    if (tcMap.TryGetProperty("id", out var idEl))
                    {
                        var id = idEl.GetString() ?? "";
                        if (!string.IsNullOrEmpty(id) && !tcState.Started)
                        {
                            tcState.Started = true;
                            tcState.ID = id;
                            var fnName = "";
                            if (tcMap.TryGetProperty("function", out var fnEl) && fnEl.TryGetProperty("name", out var nameEl))
                            {
                                fnName = nameEl.GetString() ?? "";
                                tcState.Name = fnName;
                            }

                            var outputIdx = ComputeToolOutputIndex(tcIdx);
                            events.Add(ResponsesSSE("response.output_item.added", new Dictionary<string, object>
                            {
                                ["type"] = "response.output_item.added",
                                ["output_index"] = outputIdx,
                                ["item"] = new Dictionary<string, object>
                                {
                                    ["type"] = "function_call",
                                    ["id"] = id,
                                    ["call_id"] = id,
                                    ["name"] = fnName,
                                    ["status"] = "in_progress",
                                },
                            }));
                        }
                    }

                    // Function arguments delta
                    if (tcMap.TryGetProperty("function", out var fnEl2))
                    {
                        if (fnEl2.TryGetProperty("arguments", out var argsEl))
                        {
                            var args = argsEl.GetString() ?? "";
                            if (!string.IsNullOrEmpty(args))
                            {
                                tcState.Arguments += args;
                                if (tcState.Started)
                                {
                                    var outputIdx = ComputeToolOutputIndex(tcIdx);
                                    events.Add(ResponsesSSE("response.function_call_arguments.delta", new Dictionary<string, object>
                                    {
                                        ["type"] = "response.function_call_arguments.delta",
                                        ["output_index"] = outputIdx,
                                        ["item_id"] = tcState.ID,
                                        ["delta"] = args,
                                    }));
                                }
                            }
                        }
                        // Also extract name (may appear in subsequent chunks)
                        if (fnEl2.TryGetProperty("name", out var nameEl2))
                        {
                            var name = nameEl2.GetString() ?? "";
                            if (!string.IsNullOrEmpty(name) && string.IsNullOrEmpty(tcState.Name))
                                tcState.Name = name;
                        }
                    }
                }
            }

            // ── finish_reason ──

            if (choice.TryGetProperty("finish_reason", out var frEl))
            {
                var fr = frEl.GetString() ?? "";
                if (!string.IsNullOrEmpty(fr) && !_finished)
                    _finishReason = fr;
            }
        }

        return events;
    }

    // ═══ Close ═══

    /// <summary>Close the stream (on end or error), returning remaining events.</summary>
    public List<string> Close()
    {
        if (_finished) return [];
        _finished = true;

        var events = new List<string>();

        if (_started)
        {
            events.AddRange(CloseTextBlock());
            events.AddRange(CloseToolCallBlocks());
        }
        else
        {
            events.AddRange(EmitResponseCreated());
        }

        events.Add(ResponsesSSE("response.completed", new Dictionary<string, object>
        {
            ["type"] = "response.completed",
            ["response"] = BuildResponsesSSEObject("incomplete", null),
        }));

        return events;
    }

    // ═══ Error Output ═══

    /// <summary>Generate Responses API format SSE error.</summary>
    public static string SseError(string message)
    {
        var msg = EscapeForSSE(message);
        return $"event: error\ndata: {{\"type\":\"error\",\"error\":{{\"type\":\"server_error\",\"message\":\"{msg}\"}}}}\n\n";
    }

    /// <summary>Generate Responses API format SSE context limit error.</summary>
    public static string SseContextLimitError(string message)
    {
        var msg = EscapeForSSE("request too large: " + message);
        return $"event: error\ndata: {{\"type\":\"error\",\"error\":{{\"type\":\"invalid_request_error\",\"message\":\"{msg}\"}}}}\n\n";
    }

    // ═══ Private Event Emitters ═══

    private List<string> EmitResponseCreated()
    {
        var createdResp = BuildResponsesSSEObject("in_progress", null);
        return
        [
            ResponsesSSE("response.created", new Dictionary<string, object>
            {
                ["type"] = "response.created",
                ["response"] = createdResp,
            }),
            ResponsesSSE("response.in_progress", new Dictionary<string, object>
            {
                ["type"] = "response.in_progress",
                ["response"] = createdResp,
            }),
        ];
    }

    private List<string> EmitTextStartIfNeeded()
    {
        if (_textStarted) return [];
        _textStarted = true;

        return
        [
            ResponsesSSE("response.output_item.added", new Dictionary<string, object>
            {
                ["type"] = "response.output_item.added",
                ["output_index"] = 0,
                ["item"] = new Dictionary<string, object>
                {
                    ["type"] = "message",
                    ["id"] = _outputItemID,
                    ["status"] = "in_progress",
                    ["role"] = "assistant",
                    ["content"] = new List<object>(),
                },
            }),
            ResponsesSSE("response.content_part.added", new Dictionary<string, object>
            {
                ["type"] = "response.content_part.added",
                ["output_index"] = 0,
                ["content_index"] = _contentIndex,
                ["part"] = new Dictionary<string, string> { ["type"] = "output_text", ["text"] = "" },
            }),
        ];
    }

    private List<string> EmitFinish()
    {
        var events = new List<string>();
        events.AddRange(CloseTextBlock());
        events.AddRange(CloseToolCallBlocks());
        events.AddRange(EmitCompleted());
        return events;
    }

    private List<string> CloseTextBlock()
    {
        if (!_textStarted) return [];

        return
        [
            ResponsesSSE("response.output_text.done", new Dictionary<string, object>
            {
                ["type"] = "response.output_text.done",
                ["output_index"] = 0,
                ["content_index"] = _contentIndex,
                ["text"] = _fullContent,
            }),
            ResponsesSSE("response.content_part.done", new Dictionary<string, object>
            {
                ["type"] = "response.content_part.done",
                ["output_index"] = 0,
                ["content_index"] = _contentIndex,
                ["part"] = new Dictionary<string, object>
                {
                    ["type"] = "output_text",
                    ["text"] = _fullContent,
                    ["annotations"] = new List<object>(),
                },
            }),
            ResponsesSSE("response.output_item.done", new Dictionary<string, object>
            {
                ["type"] = "response.output_item.done",
                ["output_index"] = 0,
                ["item"] = new Dictionary<string, object>
                {
                    ["type"] = "message",
                    ["id"] = _outputItemID,
                    ["status"] = "completed",
                    ["role"] = "assistant",
                    ["content"] = new List<object>
                    {
                        new Dictionary<string, object>
                        {
                            ["type"] = "output_text",
                            ["text"] = _fullContent,
                            ["annotations"] = new List<object>(),
                        },
                    },
                },
            }),
        ];
    }

    private List<string> CloseToolCallBlocks()
    {
        var events = new List<string>();
        foreach (var tcIdx in _toolCallOrder)
        {
            var tc = _toolCalls.GetValueOrDefault(tcIdx);
            if (tc == null || !tc.Started) continue;
            var outputIdx = ComputeToolOutputIndex(tcIdx);

            events.Add(ResponsesSSE("response.function_call_arguments.done", new Dictionary<string, object>
            {
                ["type"] = "response.function_call_arguments.done",
                ["output_index"] = outputIdx,
                ["item_id"] = tc.ID,
                ["arguments"] = tc.Arguments,
            }));
            events.Add(ResponsesSSE("response.output_item.done", new Dictionary<string, object>
            {
                ["type"] = "response.output_item.done",
                ["output_index"] = outputIdx,
                ["item"] = new Dictionary<string, object>
                {
                    ["type"] = "function_call",
                    ["id"] = tc.ID,
                    ["call_id"] = tc.ID,
                    ["name"] = tc.Name,
                    ["arguments"] = tc.Arguments,
                    ["status"] = "completed",
                },
            }));
        }
        return events;
    }

    private List<string> EmitCompleted()
    {
        var status = _finishReason == "length" ? "incomplete" : "completed";

        var outputItems = new List<Dictionary<string, object>>();
        if (_textStarted)
        {
            outputItems.Add(new Dictionary<string, object>
            {
                ["type"] = "message",
                ["id"] = _outputItemID,
                ["status"] = "completed",
                ["role"] = "assistant",
                ["content"] = new List<object>
                {
                    new Dictionary<string, object>
                    {
                        ["type"] = "output_text",
                        ["text"] = _fullContent,
                        ["annotations"] = new List<object>(),
                    },
                },
            });
        }

        foreach (var tcIdx in _toolCallOrder)
        {
            var tc = _toolCalls.GetValueOrDefault(tcIdx);
            if (tc == null) continue;
            outputItems.Add(new Dictionary<string, object>
            {
                ["type"] = "function_call",
                ["id"] = tc.ID,
                ["call_id"] = tc.ID,
                ["name"] = tc.Name,
                ["arguments"] = tc.Arguments,
                ["status"] = "completed",
            });
        }

        var completedResp = BuildResponsesSSEObject(status, outputItems);
        return
        [
            ResponsesSSE("response.completed", new Dictionary<string, object>
            {
                ["type"] = "response.completed",
                ["response"] = completedResp,
            }),
        ];
    }

    // ═══ Private Helpers ═══

    private int ComputeToolOutputIndex(int tcIdx)
    {
        var @base = _textStarted ? 1 : 0;
        return @base + IndexOfInt(_toolCallOrder, tcIdx);
    }

    private Dictionary<string, object> BuildResponsesSSEObject(string status, List<Dictionary<string, object>>? outputItems)
    {
        return new Dictionary<string, object>
        {
            ["id"] = _respID,
            ["object"] = "response",
            ["created_at"] = (int)DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            ["model"] = _requestedModel,
            ["status"] = status,
            ["output"] = outputItems ?? new List<Dictionary<string, object>>(),
            ["usage"] = new Dictionary<string, int>
            {
                ["input_tokens"] = _promptTokens,
                ["output_tokens"] = _completionTokens,
                ["total_tokens"] = _totalTokens > 0 ? _totalTokens : _promptTokens + _completionTokens,
            },
            ["metadata"] = new Dictionary<string, object>(),
        };
    }

    private string ResponsesSSE(string eventType, Dictionary<string, object> data)
    {
        var json = JsonSerializer.Serialize(data);
        return $"event: {eventType}\ndata: {json}\n\n";
    }

    private static (string Data, bool Done, bool Ok) ParseSSELine(string line)
    {
        var dataStr = line;
        if (line.StartsWith("data: "))
            dataStr = line[6..];
        else if (line.StartsWith("data:"))
            dataStr = line[5..];
        else
            return ("", false, false);

        dataStr = dataStr.Trim();
        if (dataStr == "[DONE]") return ("", true, true);
        if (string.IsNullOrEmpty(dataStr)) return ("", false, false);
        return (dataStr, false, true);
    }

    private static int IntFromMap(JsonElement map, string key)
    {
        if (!map.TryGetProperty(key, out var el)) return 0;
        if (el.ValueKind == JsonValueKind.Number) return el.GetInt32();
        return 0;
    }

    private static int GetInt(JsonElement el)
    {
        if (el.ValueKind == JsonValueKind.Number) return el.GetInt32();
        return 0;
    }

    private static int IndexOfInt(List<int> list, int value)
    {
        for (var i = 0; i < list.Count; i++)
            if (list[i] == value) return i;
        return -1;
    }

    private static string RandomHex(int n)
    {
        var bytes = new byte[n];
        Random.Shared.NextBytes(bytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static string EscapeForSSE(string s) =>
        s.Replace("\\", "\\\\").Replace("\"", "\\\"").Replace("\n", "\\n");
}
